### PR TITLE
f-checkout@0.156.0 - Low value order experiment headers in requests

### PR DESCRIPTION
**Pull request by charliearlie**
### Added
- Experiment headers to PATCH checkout request
- Track experiments if returned in response headers

### CODE REVIEW

File: packages/components/organisms/f-checkout/src/components/Checkout.vue [hunk 1]
```diff
@@ -299,6 +300,11 @@ export default {
         getGeoLocationUrl: {
             type: String,
             required: true
+        },
+
+        experiments: {
+            type: Object,
+            required: true
```
Comment (xander-marjoram) [inline packages/components/organisms/f-checkout/src/components/Checkout.vue:313]: I think this is fine for now but I'm not sure this needs to be required, we could probably just have it default to an empty object.

Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/components/Checkout.vue:313]: They were the two things I was torn between. I will update to add a default.

File: packages/components/organisms/f-checkout/src/services/experimentService.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// Can we have this in a public repo?
```
Comment (xander-marjoram) [inline packages/components/organisms/f-checkout/src/services/experimentService.js]: Yes, this is all client-side anyway, if someone wanted to see this they could just look in the data layer.

Comment (xander-marjoram) [reply packages/components/organisms/f-checkout/src/services/experimentService.js]: And if not, it's too late anyway 😅

Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/services/experimentService.js]: Homer hedge

File: packages/components/organisms/f-checkout/src/services/experimentService.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        return {
+            event: 'trackExperimentV2',
+            custom: {
+                experiment: {
+                    id: 'EX-1862',
+                    name: 'low_value_order_threshold',
+                    platform: 'experiment_api',
+                    version: 1,
+                    variant: {
+                        name: variant ?? 'not_applied'
```
Comment (minrwhite) [inline packages/components/organisms/f-checkout/src/services/experimentService.js]: Given this object is wrapped in an `if (variant) {` block - would this ever be set to `'not_applied'`?

File: packages/components/organisms/f-checkout/CHANGELOG.md [hunk 4]
```diff
@@ -3,13 +3,24 @@
...
+
+
 v0.151.0
 ------------------------------
 *July 5, 2021*
 
 ### Fixed
 - Get new `availableFulfilment` times when no available fulfilment times are available.
 
+
```
Comment (xander-marjoram) [inline packages/components/organisms/f-checkout/CHANGELOG.md]: Super nit: This line isn't necessary

File: packages/components/organisms/f-checkout/src/services/tests/experimentService.test.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+                // Act
+                const actual = experimentService.getLowValueOrderExperimentTracking('true');
+
+                // Assert
+                expect(actual).toEqual(expectedTrackingObject);
+            });
+        });
+        describe('when a variant is not present', () => {
+            it('should return empty address with postcode set, when postcode does not match', () => {
```
Comment (xander-marjoram) [inline packages/components/organisms/f-checkout/src/services/tests/experimentService.test.js]: This description doesn't look right to me

File: packages/components/organisms/f-checkout/src/store/checkout.module.js [hunk 6]
```diff
@@ -159,23 +159,26 @@ export default {
          * @param {Object} payload - Parameter with the different configurations for the request.
          */
         // eslint-disable-next-line no-unused-vars
-        updateCheckout: async ({ commit, state, dispatch }, {
-            url, data, timeout
-        }) => {
+        updateCheckout: async ({
+            commit, state, dispatch, rootGetters
+        }, {
+                url, data, timeout
```
Comment (xander-marjoram) [inline packages/components/organisms/f-checkout/src/store/checkout.module.js]: Nit: You can probably combine this line with the lines above and below

File: packages/components/organisms/f-checkout/src/store/tests/checkout.module.test.js [hunk 7]
```diff
@@ -348,11 +348,13 @@ describe('CheckoutModule', () => {
         let commit;
         let dispatch;
         let payload;
+        let rootGetters;
 
         beforeEach(() => {
             commit = jest.fn();
             dispatch = jest.fn();
             state = defaultState;
+            rootGetters = {};
```
Comment (minrwhite) [inline packages/components/organisms/f-checkout/src/store/tests/checkout.module.test.js]: To make the action invocations look a bit neater, these could be wrapped up into a `context` object, e.g.:

```suggestion
        let commit;
        let dispatch;
        let payload;
        let rootGetters;
        let context;

        beforeEach(() => {
            commit = jest.fn();
            dispatch = jest.fn();
            state = defaultState;
            rootGetters = {};
            context = { commit, state, dispatch, rootGetters };
```

File: packages/components/organisms/f-checkout/CHANGELOG.md (hunk 4 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/CHANGELOG.md]: I thought we kept 2 lines between each addition to the changelog so spotted that and added it 😅

File: packages/components/organisms/f-checkout/src/store/checkout.module.js (hunk 6 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/store/checkout.module.js]: Prettier did this 🙁

File: packages/components/organisms/f-checkout/src/services/tests/experimentService.test.js (hunk 5 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/services/tests/experimentService.test.js]: You are correct. I lazily copied the whole address service tests to populate this file to build upon and left that!

**Review by minrwhite — COMMENTED** [review]
A couple of comments

File: packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+import {
+    UPDATE_EXPERIMENTS_STATE
+} from './mutation-types';
+
+export default {
+    namespaced: true,
+
+    state: () => ({
+        lowValueOrderExperimentValue: ''
```
Comment (xander-marjoram) [inline packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js]: I'm wondering if having "experiment value" on the end of the variable names is necessary, I think they're probably both implied?

File: packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js [hunk 9]
```diff
@@ -402,6 +403,47 @@ describe('CheckoutAnalyticsModule', () => {
...
+                };
+
+                // Act
+                trackLowValueOrderExperiment(mockedResponseHeaders);
+
+                // Assert
+                expect(window.dataLayer).toContainEqual(expected);
+            });
+
+            it('should not `push` low value order event to data layer if it is returned in request header`', () => {
```
Comment (xander-marjoram) [inline packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js]: Should this be "if it is *not* returned in request header"?

File: packages/components/organisms/f-checkout/src/services/experimentService.js (hunk 3 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/services/experimentService.js]: Good point. I may actually be okay removing the if check then.

File: packages/components/organisms/f-checkout/CHANGELOG.md (hunk 4 above)
Comment (xander-marjoram) [reply packages/components/organisms/f-checkout/CHANGELOG.md]: We do, but only between separate version entries, not within them 😄

File: packages/components/organisms/f-checkout/src/store/checkout.module.js (hunk 6 above)
Comment (xander-marjoram) [reply packages/components/organisms/f-checkout/src/store/checkout.module.js]: Ok that's fine, I can agree to disagree with Prettier

File: packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js (hunk 8 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js]: I hijacked this from menuweb but happy to remove the suffixes.

File: packages/components/organisms/f-checkout/CHANGELOG.md (hunk 4 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/CHANGELOG.md]: Shows how much of a state I am that I didn't even notice that 😂  - updated

File: packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js (hunk 8 above)
Comment (DamianMullins) [reply packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js]: We use an `isAbcExperimentActive` convention on menu which reads nicely in components.

Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js]: I remember that now. I must have forgot what part I copied 😂

Comment (xander-marjoram) [reply packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js]: I'm gonna change it to `lowValueOrderExperimentVariant`, as that's what it seems to be being used as

File: packages/components/organisms/f-checkout/src/store/checkoutAnalytics.module.js [hunk 10]
```diff
@@ -125,6 +120,14 @@ export default {
                     dispatch('trackFormInteraction', { action: 'error', error: mappedError });
                 }
             });
+        },
+
+        trackLowValueOrderExperiment: experimentHeaders => {
```
Comment (matthewhardern) [inline packages/components/organisms/f-checkout/src/store/checkoutAnalytics.module.js:128]: can we have a comment on the action

File: packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+export default {
+    namespaced: true,
+
+    state: () => ({
+        lowValueOrderExperimentVariant: ''
+    }),
+
+    actions: {
+        setExperimentValues: ({ commit }, experiments) => {
```
Comment (matthewhardern) [inline packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js:17]: same as above

File: packages/components/organisms/f-checkout/src/store/checkoutAnalytics.module.js (hunk 10 above)
Comment (xander-marjoram) [reply packages/components/organisms/f-checkout/src/store/checkoutAnalytics.module.js:128]: Done

File: packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js (hunk 11 above)
Comment (xander-marjoram) [reply packages/components/organisms/f-checkout/src/store/checkoutExperimentation.module.js:17]: Done

**Review by matthewhardern — DISMISSED** [review]

File: packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js [hunk 12]
```diff
@@ -402,6 +402,64 @@ describe('CheckoutAnalyticsModule', () => {
...
+            it('should not `push` low value order event to data layer if it is not returned in request header`', () => {
+                // Arrange
+                const expected = {
+                    custom: {
+                        experiment: {
+                            id: 'EX-1862',
+                            name: 'low_value_order_threshold',
+                            platform: 'experiment_api',
+                            variant: {
+                                name: 'not_applied'
```
Comment (gareththackeray) [inline packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js]: I think this should be `'reserve'`.  `'not_applied'` means the user was in a bucket but it wasn't relevant to apply the experiment to them for some other reason.

File: packages/components/organisms/f-checkout/src/components/_tests/helpers/setup.js [hunk 13]
```diff
@@ -1,5 +1,7 @@
 import Vuex from 'vuex';
-import { CHECKOUT_METHOD_DELIVERY, VUEX_CHECKOUT_MODULE, VUEX_CHECKOUT_ANALYTICS_MODULE } from '../../../constants';
+import {
+    CHECKOUT_METHOD_DELIVERY, VUEX_CHECKOUT_MODULE, VUEX_CHECKOUT_ANALYTICS_MODULE, VUEX_CHECKOUT_EXPERIMENTATION_MODULE
```
Comment (kevinrodrigues) [inline packages/components/organisms/f-checkout/src/components/_tests/helpers/setup.js]: Split over single lines?

File: packages/components/organisms/f-checkout/src/store/checkoutAnalytics.module.js [hunk 14]
```diff
@@ -125,6 +120,17 @@ export default {
...
+        },
+
+        /**
+         * Fetches the variant of the Low Value Order experiment from the headers and pushes an analytics event.
+         */
+        trackLowValueOrderExperiment: experimentHeaders => {
+            const lowValueOrderExperimentVariant = experimentHeaders[HEADER_LOW_VALUE_ORDER_EXPERIMENT];
+            const event = experimentService.getLowValueOrderExperimentTracking(lowValueOrderExperimentVariant);
+            if (event) {
+                window.dataLayer.push(event);
```
Comment (kevinrodrigues) [inline packages/components/organisms/f-checkout/src/store/checkoutAnalytics.module.js:132]: Do we need to check that the dataLayer exists at all?

File: packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js [hunk 15]
```diff
@@ -402,6 +402,64 @@ describe('CheckoutAnalyticsModule', () => {
...
+                };
+
+                // Act
+                trackLowValueOrderExperiment(mockedResponseHeaders);
+
+                // Assert
+                expect(window.dataLayer).toContainEqual(expected);
+            });
+
+            it('should not `push` low value order event to data layer if it is not returned in request header`', () => {
```
Comment (kevinrodrigues) [inline packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js]: Minor: The backtick at the end of this IT statement should be a normal single quote right?

File: packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js [hunk 16]
```diff
@@ -402,6 +402,64 @@ describe('CheckoutAnalyticsModule', () => {
                 expect(window.dataLayer).toContainEqual(expected);
             });
         });
+
+        describe('trackLowValueOrderExperiment ::', () => {
+            it('should `push` low value order event to data layer if it is returned in request header`', () => {
```
Comment (kevinrodrigues) [inline packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js]: Same, backtick

**Review by kevinrodrigues — DISMISSED** [review]
Minor comments but happy to approve it.

File: packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js (hunk 16 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/store/tests/checkoutAnalytics.module.test.js]: 🦅  eagle eyed

File: packages/components/organisms/f-checkout/src/components/_tests/helpers/setup.js (hunk 13 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/components/_tests/helpers/setup.js]: Prettier did this 👀

File: packages/components/organisms/f-checkout/src/store/checkoutAnalytics.module.js (hunk 14 above)
Comment (charliearlie) [reply packages/components/organisms/f-checkout/src/store/checkoutAnalytics.module.js:132]: There are other functions above which don't check but happy to add it

**Comment by charliearlie** [comment]
> Minor comments but happy to approve it.

Fixed

**Review by kevinrodrigues — APPROVED** [review]

**Review by minrwhite — APPROVED** [review]

**Pull request merged**